### PR TITLE
Feature/machine live data

### PR DESCRIPTION
**Pull request by oxxr**

### CODE REVIEW

File: src/web/src/app/shared/components/select-machine/select-machine.component.css [hunk 1]
```diff
@@ -1,4 +1,5 @@
 .selectmachine-ngselect.dropdown {
     background-color: white;
+    margin-bottom: 1.5rem;
```
[the remark this answers is not in the document]
Comment (oxxr) [reply src/web/src/app/shared/components/select-machine/select-machine.component.css]: Yes that's correct, closing this for now but will followed in the future

File: src/web/Scenarios/MachineLiveData/MachineLiveDataController.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            _hub = hub ?? throw new ArgumentNullException(nameof(hub));
+            _machineLiveDataService = machineLiveDataService ?? throw new ArgumentNullException(nameof(machineLiveDataService));
+            _machineLiveDataService.SetCallback(SendAsync);
+        }
+
+        [HttpGet]
+        public async Task<HttpResponseMessage> Get()
+        {
+            try
```
Comment (SebastianRoether) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataController.cs]: Does this (still) serve a purpose?

File: src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        IEventProcessorHostInterface CreateEventProcessorHost();
+    }
+
+    public interface IEventProcessorHostInterface
+    {
+        Task RegisterEventProcessorFactoryAsync(IEventProcessorFactory factory, EventProcessorOptions options);
+        Task UnregisterEventProcessorAsync();
+    }
+
+    public class EventProcessorHostWrapper : IEventProcessorHostInterface
```
Comment (SebastianRoether) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs]: The wrapper is for testing purposes right? Please add a comment to make it more obvious what's going on

File: src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        }
+    }
+
+    public interface IMachineLiveDataEventProcessorFactory : IEventProcessorFactory
+    {
+        void SetCallback(Func<string, Task> func);
+        IEventProcessorHostInterface CreateEventProcessorHost();
+    }
+
+    public interface IEventProcessorHostInterface
```
Comment (SebastianRoether) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs]: `I...Interface` ? Why not just `IEventProcessorHost`?

File: src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        public Task CloseAsync(PartitionContext context, CloseReason reason)
+        {
+            return Task.CompletedTask;
+        }
+
+        public Task ProcessErrorAsync(PartitionContext context, Exception error)
+        {
+            Trace.TraceError(error.Message);
```
Comment (SebastianRoether) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs]: Don't we have a logging library?

File: src/web/Startup.cs [hunk 6]
```diff
@@ -1,120 +1,133 @@
...
+            services
+                .AddLogging(builder =>
+                {
+                    builder.AddSerilog(dispose: true);
+                })
+                .AddSingleton<IScenarioCrawler, ScenarioCrawler>()
+                .AddSingleton<IScenarioRepository, ScenarioRepository>()
+                .AddSingleton<ITokenProvider, TokenProvider>()
+                .AddSingleton<OptionsValidator>()
+                .AddSingleton<IMachineLiveDataService, MachineLiveDataService>()
```
Comment (SebastianRoether) [inline src/web/Startup.cs]: This service needs to be running all the time, not only be created the first time the controller is used

File: src/web/Scenarios/MachineLiveData/MachineLiveDataController.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            catch (Exception e)
+            {
+                Trace.TraceError(e.Message);
+                return new HttpResponseMessage(HttpStatusCode.InternalServerError);
+            }
+        }
+
+        private async Task SendAsync(string machineId, MachineLiveDataContainer data)
+        {
+            await _hub.Clients.Group(machineId).SendAsync("streamMachineData", data);
```
Comment (SebastianRoether) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataController.cs]: This setup doesn't need to be in the controller. Whenever a message is received for a machine, it should be posted in the machine group in the hub

File: src/web/Scenarios/MachineLiveData/MachineLiveDataController.cs (hunk 2 above)
Comment (oxxr) [reply src/web/Scenarios/MachineLiveData/MachineLiveDataController.cs]: Should be done

File: src/web/Startup.cs (hunk 6 above)
Comment (oxxr) [reply src/web/Startup.cs]: Should be loaded on startup -> Startup.cs -> Configure

File: src/web/Scenarios/MachineLiveData/MachineLiveDataController.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+using Microsoft.AspNetCore.SignalR;
+
+namespace Aitgmbh.Tapio.Developerapp.Web.Scenarios.MachineLiveData
+{
+    [Scenario("machine-live-data-scenario", "Machine Live Data", "/scenario-machinelivedata",
+        "src/web/src/app/scenario-machinelivedata", "src/web/Scenarios/MachineLiveData", "https://developer.tapio.one/docs/TapioDataCategories.html#streaming-data")]
+    [Produces("application/json")]
+    [Route("api/[controller]")]
+    public class MachineLiveDataController : Controller
+    {
```
Comment (FreyJim) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataController.cs]: leave a comment why this controller is empty

File: src/web/Scenarios/MachineLiveData/MachineLiveDataService.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            if (_callback != null)
+            {
+                await _callback(result.MachineId, result);
+            }
+        }
+    }
+
+    public interface IMachineLiveDataService
+    {
+        Task RegisterHubAsync();
```
Comment (FreyJim) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataService.cs]: add comments to the methods at least in a interface

File: src/web/Scenarios/MachineLiveData/MachineLiveDataCommunicationService.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        public void RegisterCallback()
+        {
+            _machineLiveDataService.SetCallback(SendAsync);
+
+        }
+        private async Task SendAsync(string machineId, MachineLiveDataContainer data)
+        {
+            await _hub.Clients.Group(machineId).SendAsync("streamMachineData", data);
```
Comment (FreyJim) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataCommunicationService.cs]: please add an `.ConfigureAwait(false)` to async methods

File: src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+                PartitionReceiver.DefaultConsumerGroupName,
+                credentialProvider.GetEventHubConnectionString(),
+                credentialProvider.GetStorageConnectionString(),
+                credentialProvider.GetStorageContainerName()
+            );
+        }
+
+        public async Task RegisterEventProcessorFactoryAsync(IEventProcessorFactory factory, EventProcessorOptions options)
+        {
+            await _wrappedProcessorHost.RegisterEventProcessorFactoryAsync(factory, options);
```
Comment (FreyJim) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs]: please add an `.ConfigureAwait(false)` to async methods

File: src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        }
+
+        public async Task UnregisterEventProcessorAsync()
+        {
+            await _wrappedProcessorHost.UnregisterEventProcessorAsync();
```
Comment (FreyJim) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataEventProcessorFactory.cs]: please add an `.ConfigureAwait(false)` to async methods

File: src/web/Scenarios/MachineLiveData/MachineLiveDataHub.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+        public async Task LeaveGroup(string groupName)
+        {
+            await Groups.RemoveFromGroupAsync(Context.ConnectionId, groupName);
+        }
+
+        public override async Task OnConnectedAsync()
+        {
+            MachineLiveDataConnectionHandler.Connections.Add(Context.ConnectionId);
+            await base.OnConnectedAsync();
```
Comment (FreyJim) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataHub.cs]: please add an `.ConfigureAwait(false)` to async methods

File: src/web/Scenarios/MachineLiveData/MachineLiveDataHub.cs [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        public override async Task OnDisconnectedAsync(Exception exception)
+        {
+            MachineLiveDataConnectionHandler.Connections.Remove(Context.ConnectionId);
+            await base.OnDisconnectedAsync(exception);
```
Comment (FreyJim) [inline src/web/Scenarios/MachineLiveData/MachineLiveDataHub.cs]: please add an `.ConfigureAwait(false)` to async methods

**Pull request merged**